### ISSUE TITLE
os.kill in twisted.internet.process has no error handling

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2420]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#2420|
|Type|defect|
|Created|2007-01-25 02:57:22Z|
|Branch|https://github.com/twisted/twisted/tree/process-kill-2420-2|

It may be that our code will never trigger os.kill complaining about non-existent process, but we should handle it anyway.

<details><summary>Searchable metadata</summary>

```
trac-id__2420 2420
type__defect defect
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__branches_process_kill_2420_2 branches/process-kill-2420-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1169693842000000 1169693842000000
changetime__1341772809000000 1341772809000000
version__None None
owner__therve therve
cc__therve cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

Implementation of error handling needs to be moved out of [#2341](https://github.com/twisted/twisted/issues/2341), and then tests need to be added.

Any interest in doing this, therve?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

Of course :).

But if we don't wait for [#2341](https://github.com/twisted/twisted/issues/2341), we'll have to duplicate the code here, no ?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @therve|
|-|-|

After discussion, we'll wait for [#2341](https://github.com/twisted/twisted/issues/2341).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

[#2787](https://github.com/twisted/twisted/issues/2787) is related probably.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [23848]) Branching to 'process-kill-2420'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [23849]) Handle ESRCH in os.kill.

Refs [#2420](https://github.com/twisted/twisted/issues/2420)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

  * the way `test_killExited` relies on the call to `os.waitpid` in `registerReapProcessHandler` is pretty obscure (I couldn't figure it out, you had to explain it to me on IRC :).  Some clarification would be great.
  * I think the docstring for `test_killExitedButNotDetected` is slightly misleading.  If a process exits but it hasn't been waited on yet (ie, it is a zombie) then `os.kill` will not raise an `ESRCH`  error.  The kill will succeed.  You can only get `ESRCH` after the zombie is reaped.  This is related to what has always been the problem with this ticket - no one knows how to trigger the `ESRCH` case in the real world. :P  That makes it hard to describe the case `test_killExitedButNotDetected` is testing.  Maybe the thing that could really cause this is if some other library/application code calls `wait` and gets a process Twisted started.  So maybe describe it that way, instead?  Of course, this case has a bunch of other bugs - like the fact that the process will never get `processEnded` notification, so fixing `signalProcess` only half fixes it.
  * Another errno that seems more likely (though I've never seen it, so who knows) to happen is `EPERM`.  At least I can imagine a legitimate use of the API which would result in this (eg, start a process as root, spawn a child, both processes then change uid/gid to values different from each other, then `signalProcess` is attempted - permission denied).  It should be easy to test/handle this case as well, at least for POSIX.  I suspect something similar can happen on windows if someone plays around with permissions enough, so it might be good to have a uniform exception for this case (or I might be wrong and they might not be similar enough for that, I dunno).  This can be a separate ticket if you want, I just mention it because it's pretty related to these changes, and because it will slightly invalidate the docstring of `test_killErrorInKill`, although not the test code itself.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [24162]) Add a comment, fix docstring.

Refs [#2420](https://github.com/twisted/twisted/issues/2420)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

Hey Thomas, you should fix this. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [30368]) Branching to 'process-kill-2420-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

OK, it's ready to review I think. I didn't do anything about `EPERM`, we can open another ticket if it's considered important.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

  1. Needs a news fragment
  1. Some documentation somewhere needs to be updated, I guess, to explain that `spawnProcess` is allowed to fail with `OSError` sometimes (but not in the `ESRCH` case).
  1. It would be nice if a similar test/fix could be done to the Windows process implementation.  But this could be a separate ticket.
  1. It would also be nice if this could be a test in `twisted/internet/test/`.  I can see how that might challenging since no one knows how to trigger this case for real. :)  Maybe for EPERM something is possible.  This could be a separate ticket too.

Address to your satisfaction and then merge.  Thanks!

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set owner to @therve|
|-|-|

Replying to [exarkun](#note_14):
> Address to your satisfaction and then merge.  Thanks!

I think this got lost, assigning it back to therve (of course feel free to unassign again).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `closed`|
|-|-|

(In [34700]) Merge process-kill-2420-2

Author: therve
Reviewer: exarkun
Fixes: [#2420](https://github.com/twisted/twisted/issues/2420)

Handle the ESRCH errno possibly raised by os.kill in Process.signalProcess and
raise ProcessExitedAlready instead.
